### PR TITLE
go extension filter state get api

### PR DESCRIPTION
**Pull request by StarryVae**
Commit Message: go extension filter state get api
Additional Description: N/A
Risk Level: low
Testing: intergation test
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
Go extension filterstate get api is implemented refer to #26883 , PTAL, thanks. @doujiang24

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 1]
```diff
@@ -1106,6 +1106,54 @@ CAPIStatus Filter::setStringFilterState(absl::string_view key, absl::string_view
...
+
+  if (state.isThreadSafe()) {
+    auto go_filter_state =
+        state.streamInfo().filterState()->getDataReadOnly<GoStringFilterState>(key);
+    if (go_filter_state) {
+      req_->strValue = go_filter_state->value();
+      value_str->p = req_->strValue.data();
+      value_str->n = req_->strValue.length();
+    }
+    dlib->envoyGoRequestSemaDec(req_);
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:1134]: same as https://github.com/envoyproxy/envoy/pull/26883#discussion_r1174704822

File: contrib/golang/filters/http/source/go/pkg/http/shim.go [hunk 2]
```diff
@@ -208,6 +209,9 @@ func envoyGoFilterOnHttpDestroy(r *C.httpRequest, reason uint64) {
 	req := getRequest(r)
 	// do nothing even when req.panic is true, since filter is already destroying.
 	defer req.RecoverPanic()
+	if atomic.LoadInt32(&req.waitingOnEnvoy) != 0 {
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/shim.go:212]: same as https://github.com/envoyproxy/envoy/pull/26883#discussion_r1174695680

Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/http/shim.go:212]: maybe we could use `CompareAndSwapInt32` before `sema.Done`, also need this protect in `envoyGoRequestSemaDec`.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 3]
```diff
@@ -1106,6 +1106,54 @@ CAPIStatus Filter::setStringFilterState(absl::string_view key, absl::string_view
...
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  auto dlib = Dso::DsoManager<Dso::HttpFilterDsoImpl>::getDsoByID(config_->soId());
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:1123]: also, please rebase main branch, we don't use this way to get dlib now.

File: contrib/golang/filters/http/source/go/pkg/http/shim.go (hunk 2 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/go/pkg/http/shim.go:212]: ok, thanks.

File: contrib/golang/filters/http/source/golang_filter.cc (hunk 3 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: done, PTAL again, thanks.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 4]
```diff
@@ -1106,6 +1106,54 @@ CAPIStatus Filter::setStringFilterState(absl::string_view key, absl::string_view
...
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  auto dlib = Dso::DsoManager<Dso::HttpFilterDsoImpl>::getDsoByPluginName(config_->pluginName());
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:1123]: we can use `dynamic_lib_` here.

**Review by doujiang24 — COMMENTED** [review]
Thanks, I think we are nearly close.

File: contrib/golang/filters/http/source/golang_filter.cc (hunk 4 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: done, thanks.

**Review by doujiang24 — COMMENTED** [review]
just a small issue, otherwise, LGTM, thanks!

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 5]
```diff
@@ -1106,6 +1106,51 @@ CAPIStatus Filter::setStringFilterState(absl::string_view key, absl::string_view
...
+      req_->strValue = go_filter_state->value();
+      value_str->p = req_->strValue.data();
+      value_str->n = req_->strValue.length();
+    }
+  } else {
+    auto key_str = std::string(key);
+    auto weak_ptr = weak_from_this();
+    state.getDispatcher().post([this, &state, weak_ptr, key_str, value_str] {
+      if (!weak_ptr.expired() && !hasDestroyed()) {
+        Thread::LockGuard lock(mutex_);
```
[the remark this answers is not in the document]
Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: yeah, done, thanks.

**Review by doujiang24 — APPROVED** [review]
lgtm, thanks!

**Comment by StarryVae** [comment]
cc @wbpcode @mattklein123

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**